stdlib/char.ml

```diff
@@ -74,3 +76,40 @@ type t = char
 
 let compare c1 c2 = code c1 - code c2
 let equal (c1: t) (c2: t) = compare c1 c2 = 0
+
+(* For bitvects: counting the number of bits of a character *)
+
+(*
+  let bitcount =
+    let bytes = Bytes.create 256 in
+    for c = 0 to 255 do
+      let nbits = ref 0 in
+      for b = 0 to 7 do
+        if (1 lsl b) land c <> 0 then incr nbits
+      done;
+      bytes.[c] <- Char.chr !nbits
+    done;
+    Bytes.unsafe_to_string bytes
+*)
+
+let bitcount =
+  "\000\001\001\002\001\002\002\003\001\002\002\003\002\003\003\004\
+   \001\002\002\003\002\003\003\004\002\003\003\004\003\004\004\005\
+   \001\002\002\003\002\003\003\004\002\003\003\004\003\004\004\005\
+   \002\003\003\004\003\004\004\005\003\004\004\005\004\005\005\006\
+   \001\002\002\003\002\003\003\004\002\003\003\004\003\004\004\005\
+   \002\003\003\004\003\004\004\005\003\004\004\005\004\005\005\006\
+   \002\003\003\004\003\004\004\005\003\004\004\005\004\005\005\006\
+   \003\004\004\005\004\005\005\006\004\005\005\006\005\006\006\007\
+   \001\002\002\003\002\003\003\004\002\003\003\004\003\004\004\005\
+   \002\003\003\004\003\004\004\005\003\004\004\005\004\005\005\006\
+   \002\003\003\004\003\004\004\005\003\004\004\005\004\005\005\006\
+   \003\004\004\005\004\005\005\006\004\005\005\006\005\006\006\007\
+   \002\003\003\004\003\004\004\005\003\004\004\005\004\005\005\006\
+   \003\004\004\005\004\005\005\006\004\005\005\006\005\006\006\007\
+   \003\004\004\005\004\005\005\006\004\005\005\006\005\006\006\007\
+   \004\005\005\006\005\006\006\007\005\006\006\007\006\007\007\b"
+
+(* Return the number of bits sets in a char *)
+let bits c =
+ code (string_unsafe_get bitcount (code c))
```

Comment:
It may be interesting to do some benchmarking in order to assess
whether a table lookup is the best implementation. Hacker's delight
is your friend here, for example [1] has multiple implementations of
population count that are neither loop- nor table-baseD.

[1] http://www.hackersdelight.org/hdcodetxt/pop.c.txt

Reply:
Alternatively, the table has size 256 bytes, which is 4 cache lines on x86. Since the max value is 8, you could fit it into 2 cache lines if you use nibbles.

Reply:
Also https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer

Reply:
(Of course, for a byte, this is even simpler.)

Reply:
Indeed, see `pop7` in [1] (2 mult, 2 shift, 1 and).